src/api/js/scripts/parse-api.js

```diff
@@ -0,0 +1,354 @@
+'use strict';
+
+let fs = require('fs');
+let path = require('path');
+
+let files = [
+  'z3_api.h',
+  'z3_algebraic.h',
+  'z3_ast_containers.h',
+  'z3_fixedpoint.h',
+  'z3_fpa.h',
+  'z3_optimization.h',
+  'z3_polynomial.h',
+  'z3_rcf.h',
+  'z3_spacer.h',
+];
+
+let aliases = {
+  __proto__: null,
+  Z3_bool: 'boolean',
+  Z3_string: 'string',
+  bool: 'boolean',
+  signed: 'int',
+};
+
+let primitiveTypes = {
+  __proto__: null,
+  Z3_char_ptr: 'string',
+  unsigned: 'number',
+  int: 'number',
+  uint64_t: 'bigint',
+  int64_t: 'bigint',
+  double: 'number',
+  float: 'number',
+};
+
+let optTypes = {
+  __proto__: null,
+
+  Z3_sort_opt: 'Z3_sort',
+  Z3_ast_opt: 'Z3_ast',
+  Z3_func_interp_opt: 'Z3_func_interp',
+};
+
+// parse type declarations
+let types = {
+  __proto__: null,
+
+  // these are function types I can't be bothered to parse
```

Comment:
I will likely create other such callbacks so if this list is required the more sustainable way is to mark them in z3_api as callbacks using some no-op macro. Based on the code below it looks like you match for "typedef enum" and other typedefs are not considered.

Reply:
I added macros so the closures can be discovered instead of hard-wired

Reply:
I can add support for parsing these properly, but to be sure I've done it right I'd like to have at least _a_ test of them. Right now I can't find any uses of any of these types to compare against. Can you write a brief example (in C/C++) demonstrating how any one of the consumers of the `Z3_whatever_eh` types is intended to work, such that I could tell if my implementation has the right behavior?

Reply:
https://github.com/Z3Prover/z3/blob/56d3718cde7260d54c34e1420ef11a3436671139/src/api/python/z3/z3.py#L212

Other callbacks are not exercised yet in examples I have published for python. They are used from C++
https://github.com/Z3Prover/z3/blob/56d3718cde7260d54c34e1420ef11a3436671139/src/api/c%2B%2B/z3%2B%2B.h#L3966
and then for a self-contained example:
https://github.com/Z3Prover/z3/tree/master/examples/userPropagator

Reply:
https://github.com/Z3Prover/z3/tree/master/examples/userPropagator looks like something I can re-implement, thanks. That PDF is very helpful.

For `Z3_set_error_handler`, what's an example of a case where it should be called? For example, should I expect it to be called if I perform `Z3_query_constructor` on a `Z3_constructor` which has not previously been passed to `Z3_mk_datatypes`?

Reply:
Yes, the error handler gets invoked whenever there is some mistake in the API call. 
For example, if you add two arrays. The addition is not well typed and it fails to create the expression.
Without the error handler it should return null. With the error handler you can raise an exception from the handler assuming stack unwinding works (which I guess for WebAssembly it should).

Reply:
Another question: this line:

 https://github.com/Z3Prover/z3/blob/56d3718cde7260d54c34e1420ef11a3436671139/src/api/z3_api.h#L6700

refers to a `Z3_propagate_solver_declare`. I can't find any other mention of that. Is it just a typo?